### ISSUE TITLE
Detect globals and define them as necessary.

### ISSUE THREAD

**Issue by m-mujica** [stealjs/steal#571]
If you try to use latest lodash with `steal` it will cause runtime errors when the app is bundled because `global` is undefined. Steal (steal-tools maybe?) should detect those globals and define them as necessary, browserify does something like this:

```
When opts.detectGlobals is true, scan all files for process, global, __filename, and 
__dirname, defining as necessary. With this option npm modules are more likely to 
work but bundling takes longer. Default true.
```

They also have an `insertGlobals` which always insert the globals mentioned before.

```
When opts.insertGlobals is true, always insert process, global, __filename, and 
__dirname without analyzing the AST for faster builds but larger output bundles.
Default false.
```

**Comment by matthewp** [comment]
We could add `global` here: https://github.com/stealjs/system-npm/blob/master/npm.js#L263 for everyone, I think that's fine.

__filename and __dirname are a CommonJS thing, so we *should* be adding that when transpiling to AMD, but may not be.  Good bug.

**Comment by rrs** [comment]
Is there any sort of work around for this at the moment? I have the same issue as the op.

**Comment by matthewp** [comment]
This was fixed by https://github.com/stealjs/transpile/pull/53, it should be working in the latest steal-tools 0.16.  Is it not for you?

**Comment by rrs** [comment]
Sadly no.
I am using npm. I have latest versions of all my dependencies. v0.16.19 for steal. v0.16.4 for steal-tools. I am using this build.js from an example
`stealTools.build({
  main: "js/boot",
  config: __dirname + "/package.json!npm"
}, {
  bundleAssets: true
});`

I tried excluding it from the bundle using the method at the bottom of this page http://stealjs.com/docs/load.metadata.html but then the error was replaced by another.

I'll give more detail if it will help.

**Comment by matthewp** [comment]
Yeah, what is the dependency that expects `global`?  Is it an npm package?  If so I can take a look.

**Comment by rrs** [comment]
its lodash v4.13.1 and it throws the wobbly in lodash-compat@3.10.2#internal/SetCache

**Comment by matthewp** [comment]
Thanks, I'll take a look.  One workaround you can do is add a script tag in your page like:

```
<script>
  global = window;
</script>
```

**Comment by rrs** [comment]
ahhh I tried something like that but didn't get it right. Thank you

**Comment by matthewp** [comment]
@rrs This project demonstrates importing lodash and it working in both dev and production: https://gist.github.com/matthewp/ce56d5370bd0e787a958b481093d161c

Maybe you are not importing lodash directly like `require('lodash');` but importing something else?  can you provide more info?

**Comment by rrs** [comment]
Only difference I can see between them is that i import in the es6 style
`import _ from ''lodash"`, 
I am using the current release version of steal / steal-tools not the pre.
I use a build.js file which is specified in the package.json with "build": "node build".

I will test your example tomorrow and have a play and report back.

**Comment by pYr0x** [comment]
I think the problem is that you are using es6 and the fix was only for cjs

**Comment by matthewp** [comment]
I've updated the example to import using es6, still works, that shouldn't matter.

**Comment by matthewp** [comment]
I was able to recreate with "lodash.uniq": "^3.2.2". The problem is that a dependency has some code like:

```
/** Native method references. */
var Set = getNative(global, 'Set');
```

We detect CommonJS use of `global` [here](https://github.com/stealjs/transpile/blob/c4fa843e4fc6dc06329f2df55707e591f5044957/lib/cjs_amd.js#L11) and then wrap it like:

```js
(function(global){
  // cjs code here
})(window);
```

However our detection code is too strict and expects you to use `global.foo` with the dot operator. This code isn't doing so.  

I think it's fine to make this less restrictive. Ideally we would eliminate comments and strings like steal does [here](https://github.com/stealjs/system-trace/blob/master/trace.js#L98), but even just removing the `.` check would be an improvement.

**Comment by rrs** [comment]
The bit it is failing on for me is exactly that line in lodash-compact/internal/SetCache.js

```
/** Native method references. */
var Set = getNative(global, 'Set');
```

**Comment by matthewp** [comment]
Will be fixed by https://github.com/stealjs/transpile/pull/54, releasing on Monday (Friday releases always go wrong).

**Comment by matthewp** [comment]
Released in transpile 0.9.7.
